R/expr-array.R

```diff
@@ -0,0 +1,335 @@
+#'   values = list(c("a", "b", "c"), c("x", "y", "z"), c("e", NA, NA)),
+#'   separator = c("-", "+", "/"),
+#' )$cast(values = pl$Array(pl$String, 3))
+#' df$with_columns(
+#'   join_with_expr = pl$col("values")$arr$join(pl$col("separator")),
+#'   join_with_lit = pl$col("values")$arr$join(" "),
+#'   join_ignore_null = pl$col("values")$arr$join(" ", ignore_nulls = TRUE)
+#' )
+expr_arr_join <- function(separator, ignore_nulls = FALSE) {
+  self$`_rexpr`$arr_join(as_polars_expr(separator, as_lit = TRUE)$`_rexpr`, ignore_nulls) |>
+    wrap()
+}
+
+#' Get the index of the minimal value in an array
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(1:2, 2:1)
+#' )$cast(pl$Array(pl$Int32, 2))
+#' df$with_columns(
+#'   arg_min = pl$col("values")$arr$arg_min()
+#' )
+expr_arr_arg_min <- function() {
+  self$`_rexpr`$arr_arg_min() |>
+    wrap()
+}
+
+#' Get the index of the maximal value in an array
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(1:2, 2:1)
+#' )$cast(pl$Array(pl$Int32, 2))
+#' df$with_columns(
+#'   arg_max = pl$col("values")$arr$arg_max()
+#' )
+expr_arr_arg_max <- function() {
+  self$`_rexpr`$arr_arg_max() |>
+    wrap()
+}
+
+#' Evaluate whether all boolean values in an array are true
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(c(TRUE, TRUE), c(FALSE, TRUE), c(FALSE, FALSE), c(NA, NA)),
+#' )$cast(pl$Array(pl$Boolean, 2))
+#' df$with_columns(all = pl$col("values")$arr$all())
+expr_arr_all <- function() {
+  self$`_rexpr`$arr_all() |>
+    wrap()
+}
+
+#' Evaluate whether any boolean values in an array are true
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(c(TRUE, TRUE), c(FALSE, TRUE), c(FALSE, FALSE), c(NA, NA)),
+#' )$cast(pl$Array(pl$Boolean, 2))
+#' df$with_columns(any = pl$col("values")$arr$any())
+expr_arr_any <- function() {
+  self$`_rexpr`$arr_any() |>
+    wrap()
+}
+
+#' Shift array values by `n` indices
+#'
+#' @inheritParams DataFrame_shift
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(1:3, c(2L, NA, 5L)),
+#'   idx = 1:2,
+#' )$cast(values = pl$Array(pl$Int32, 3))
+#' df$with_columns(
+#'   shift_by_expr = pl$col("values")$arr$shift(pl$col("idx")),
+#'   shift_by_lit = pl$col("values")$arr$shift(2)
+#' )
+expr_arr_shift <- function(n = 1) {
+  self$`_rexpr`$arr_shift(as_polars_expr(n, as_lit = TRUE)$`_rexpr`) |>
+    wrap()
+}
+
+
+#' Convert an Array column into a List column with the same inner data type
+#'
+#' @return [Expr][Expr_class] of [data type List][DataType_List]
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2), c(3, 4))
+#' )$cast(pl$Array(pl$Int8, 2))
+#'
+#' df$with_columns(
+#'   list = pl$col("a")$arr$to_list()
+#' )
+expr_arr_to_list <- function() {
+  self$`_rexpr`$arr_to_list() |>
+    wrap()
+}
+
+
+# TODO-REWRITE: implement this
```

Comment:
Why not `# TODO: `?
If you use your own notation, it may not be supported by commonly used tools.
I use https://github.com/Gruntfuggly/todo-tree, but it does not support this by default.

Reply:
Because some TODO come from before the rewrite, so I want to be able to distinguish those from those related to the rewrite. Would it work to have `TODO: REWRITE`? I just need something that I can search for across all docs

Reply:
Ok, thanks for clarifying.

Reply:
Thinking about it again, I don't understand what 'some TODO come from before the rewrite' means.
Can such a thing exist?

Reply:
Yes because I copy code, comments, and tests from current r-polars, and there can be existing TODO there

Reply:
I do not understand what you mean by distinguishing between them.

- If the TODO is a problem specific to r-polars and not in py-polars, it may be removed at the stage of being copied here.
  Like this https://github.com/eitsupi/neo-r-polars/pull/19#discussion_r1805532525
- If its a TODO that is not such a problem, and common problems exist here as well, why the need to distinguish?
- How do you distinguish between the TODOs I have added here from the beginning and the ones that have been copied?

Reply:
It's just that in the process of rewriting I sometimes mark some stuff as `TODO-REWRITE` because we can't run them for now because they are missing some other functions. For instance we can't run `$list$concat()` because it requires `pl$concat_list()` to be implemented first. I just want to be able quickly find those things that are related to the rewriting process only. But as I said above, if this messes up with your tool for finding TODOs, I can rename those `TODO: REWRITE` or something else, I just need a string that is unique enough to search across docs.